### PR TITLE
feat(new sink): Adding greptimedb metrics sink

### PR DESCRIPTION
**Pull request by sunng87**
This patch adds [greptimedb](https://github.com/greptimeteam/greptimedb) sink. 

For now, we use greptimedb's git repository for adding its client. We will eventually split it from repo and publish it to crates.io, I hope this is not a road blocker at the moment.

TODOs:

- [x] website docs
- [x] spellcheck cleanup
- [x] integration test verification

### CODE REVIEW

**Comment by jszwedko** [comment]
Thanks for this submission @sunng87 !

We have a few questions:

* We hadn't yet had a Vector user ask for this (in Github issues at least). Are you aware of Vector users needing this sink? Or is the hope that you'll grow GreptimeDB's community by way of adding integrations like these to other projects?
* We noticed that GreptimeDB has a MySQL-compatible interface. What are the pros/cons of using that interface vs. the native client you introduced here?
* Our team lacks familiarity with GreptimeDB. If we were to add this, would you be able to continue to support this sink in the future?

Thank you!

**Comment by sunng87** [comment]
@jszwedko Thank you for your questions.

1. GreptimeDB is relatively new and while seeking for ingestion solutions we found vector a great match. We are going to promote vector as one of our recommended agents. So it would be nice if we can upstream our sink.
2. MySQL interface works great for data analysis scenarios. But in ingestion of devops/monitoring data, it lacks support for auto table creation, which is a road blocker for us. Although prometheus remote write and influx line protocol over http are also supported by greptimedb, we hope to use a binary and column-oriented format in our recommended agent.
3. Yes I will. Our community and I will continuously support and improve this sink because it plays critical role in greptimedb's stack. Also I hope to contribute more to vector not only about the sink but also whole project.

**Comment by sunng87** [comment]
@StephenWakely I haven't fully tested those protocols via vector but I think it should work if our implementation is correct. The reason we use our own protocol is because we want a compact, controllable default ingestion protocol in our recommended agent. We can keep adding new features as our database iterates.

**Comment by sunng87** [comment]
The integration is almost done. Let me know if you find anything that requires changes.

**Comment by jszwedko** [comment]
Thanks @sunng87 ! We'll try get this another review this upcoming week.

**Comment by sunng87** [comment]
@jszwedko Thank you very much! The recent clippy error should be easy to fix. And there is an issue with container network in integration test that I will be checking next week. I hope the integration test won't block us and I can fix it together with issues from review.

File: src/sinks/greptimedb/client.rs
Comment (StephenWakely) [inline src/sinks/greptimedb/client.rs:112]: We are deprecating the use of Event sinks and are moving towards using Event Stream sinks. This will need to be restructured a little to use this.

```suggestion
        Ok(VectorSink::from_event_streamsink(sink))
```

You can find a couple of tutorials [here](https://github.com/vectordotdev/vector/tree/master/docs/tutorials/sinks) that should guide you towards this. Have a look at the [Splunk Hec Metrics sink](https://github.com/vectordotdev/vector/tree/master/src/sinks/splunk_hec/metrics) which I believe is fairly similar to what you are trying to do here.

**Review by StephenWakely — CHANGES_REQUESTED** [review]
Many thanks for submitting this. Unfortunately this is using a number of deprecated components and will need to be restructured to avoid these. I've put some links to some tutorials and a sample sink to follow.

Please reach out if you have any further questions and we'll do what we can to get this over the line.

**Comment by sunng87** [comment]
@StephenWakely Thanks for your suggestion! I have refactored the sink into `StreamSink` based APIs. There are some residual questions:

- The `RequestBuilder` and `Encode` API are bytes oriented. However, GreptimeDB's client is based on gRPC whose serialization happens within the client so we have access to bytes. In this implementation, I created a `GreptimeDBBatchSizer` to estimate the request size. `RequestBuilder` API is not used in this example, I use a simple `map` to create request from metrics.
- I didn't see `RetryLogic` applied on new APIs. How do we config retry condition and strategies?

**Comment by StephenWakely** [comment]
> * The `RequestBuilder` and `Encode` API are bytes oriented. However, GreptimeDB's client is based on gRPC whose serialization happens within the client so we have access to bytes. In this implementation, I created a `GreptimeDBBatchSizer` to estimate the request size. `RequestBuilder` API is not used in this example, I use a simple `map` to create request from metrics.

Yes, this is fine.

**Comment by sunng87** [comment]
Nice! I have finished the refactoring and made integration test working. Please take a look when you get time. Thanks!

**Review by StephenWakely — CHANGES_REQUESTED** [review]
This is looking really good! I just have a few minor changes that are needed. 

We also need a `cue` file created for the website that will pull in the generated documentation. Have a look at the [amqp sink](https://github.com/vectordotdev/vector/blob/master/website/cue/reference/components/sinks/amqp.cue) for an example of how to do this.

**Comment by sunng87** [comment]
@StephenWakely Thanks for your suggestion. Hopefully I have addressed all comments above and created website cue files for greptimedb sink.

A minor question, since we are not using vector's TLS facilities, the content generated from features.send.tls might not apply to this sink. Should I just make tls.enabled as `false`? However, greptimedb client actually uses TLS if server configured.

**Comment by StephenWakely** [comment]
> @StephenWakely Thanks for your suggestion. Hopefully I have addressed all comments above and created website cue files for greptimedb sink.
> 
> A minor question, since we are not using vector's TLS facilities, the content generated from features.send.tls might not apply to this sink. Should I just make tls.enabled as `false`? However, greptimedb client actually uses TLS if server configured.

Ah, apologies. You also need to add a file in `/website/content/en/docs/reference/configuration/sinks/greptimedb.md`. With that you should then be able to preview the documentation [here](https://deploy-preview-17198--vector-project.netlify.app/components/).

Regarding the TLS settings, a lot of users like to set custom locations for the CA certificates etc.. as the global ones are not always suitable. Is there a way you can add `pub(crate) tls: Option<crate::tls::TlsConfig>` to the configuration and pass these settings into the greptime library?

File: src/sinks/greptimedb/service.rs
Comment (StephenWakely) [inline src/sinks/greptimedb/service.rs:17]: Can you change it to:

```suggestion
use crate::sinks::prelude::*;
```

Then there are a whole bunch of other imports that can be removed. This should be done in the other sink files too.

**Comment by sunng87** [comment]
@StephenWakely Just added custom ca and client certificate support. And it's also possible to preview the docs now.

The CI is failing because this "Validate comment" CI check doesn't seem to be able to run on forked pull request, which has no access to secrets of this project.

File: src/sinks/greptimedb/service.rs
Comment (StephenWakely) [inline src/sinks/greptimedb/service.rs:139]: This shouldn't be a panic. Can you change this to return a `Result::Err`? You can use `Snafu` similar to [here](https://github.com/vectordotdev/vector/blob/master/src/nats.rs#L8-L16) to define the errors.

**Review by StephenWakely — CHANGES_REQUESTED** [review]
Nice, this is getting really close now. I just think we need to make sure it doesn't panic with incorrect TLS settings.

File: src/sinks/greptimedb/service.rs
Comment (sunng87) [reply src/sinks/greptimedb/service.rs:139]: hmmm.. I checked other sinks for `expect` usage. But no problem, let me draft an update for this.

Comment (StephenWakely) [reply src/sinks/greptimedb/service.rs:139]: Ah, if that's the case, we need to update those sinks too.

Invalid settings should result in an error message rather than a panic. This is primarily to handle the situation where a config is reloaded. If you attempt to reload an invalid config it should continue to run with the old config. A panic would kill the whole process.

Comment (sunng87) [reply src/sinks/greptimedb/service.rs]: That's reasonable. I had a quick grep and found https://github.com/vectordotdev/vector/blob/master/src/sinks/datadog/metrics/service.rs#L162 that you might want to check.

**Comment by sunng87** [comment]
Updated. @StephenWakely PTAL

File: src/sinks/greptimedb/service.rs
Comment (dsmith3197) [inline src/sinks/greptimedb/service.rs:180]: Small nit, you should be able to avoid this clone by accessing the field directly.

```suggestion
            let metadata = req.metadata;
            let result = client.insert(req.items).await?;
```

File: src/sinks/greptimedb/service.rs
Comment (dsmith3197) [inline src/sinks/greptimedb/service.rs:152]: If the `verify_certificate`, `verify_hostname` and `alpn_protocols` fields are ignored, can we add corresponding warning messages when they are set?

https://github.com/vectordotdev/vector/blob/b91976da24f75150fb800fa11deea620e4b5abbb/lib/vector-core/src/tls/settings.rs#L87

**Review by dsmith3197 — COMMENTED** [review]
Taking over from @StephenWakely. I took a look through code and everything looks good. Thank you for all the effort you put into polishing this. I've left some minor comments and suggestions that should all be straightforward.

I haven't had a chance yet to look at the documentation changes. I will hopefully have time to look through those tomorrow.

**Review by dsmith3197 — COMMENTED** [review]
[Reference docs](https://deploy-preview-17198--vector-project.netlify.app/docs/reference/configuration/sinks/greptimedb) look good to me!

We'll be ready to approve and merge once the remaining comments are addressed.

File: src/sinks/greptimedb/request_builder.rs
Comment (sunng87) [reply src/sinks/greptimedb/request_builder.rs]: Good suggestion! I'm going to extract quantiles in the next patch.

**Comment by sunng87** [comment]
@dsmith3197 Thank you for your in-depth review. I've addressed all the issues. Magic numbers have been removed by test guarded constants.

**Review by dsmith3197 — APPROVED** [review]
Looks amazing, thank you for addressing all the feedback!

**Review by StephenWakely — APPROVED** [review]
Awesome. This is looking really good now. Thanks!

**Comment by sunng87** [comment]
@StephenWakely @dsmith3197 Thank you for your help!

**Comment by StephenWakely** [comment]
> Thank you for your help!

Thank you for the contribution!

**Pull request merged**
